Snapshot store

joost-de-vries opened the pull request:
I've implemented support for snapshot storage. #56 

I had to refactor some shared logic from the `akka.persistence.dynamodb.journal` package to `akka.persistence.dynamodb`. I did not refactor the configuration properties for shared values. So currently properties for the snapshot are be default a copy of the journals values.

ktoso commented on README.md:
```diff
@@ -54,6 +55,27 @@ Before you can use these settings you will have to create a table, e.g. using th
   * a hash key of type String with name `par`
   * a sort key of type Number with name `num`
+  
+### Snapshot
```
`Snapshot Store` perhaps :)

joost-de-vries commented:
@ktoso tx for your code review. I incorporated your comment.

joost-de-vries commented:
During Scala Days @patriknw mentioned the concern that the snapshot might be too big for dynamodb. Currently there's no special snapshots that are bigger than the Dynamodb maximum item size.
I updated the readme to document the maximum size.

joost-de-vries commented:
@ktoso I'm wondering what needs to happen to merge this PR? 
I guess my first question is: who do I need to contact? Tx in advance for your help.

joost-de-vries commented:
Yeah, that's great.
As I mentioned the things I needed to address implementing snapshot support where mostly:
- how to deal with shared code
- how to deal with shared configurations

I'm actually working on implementing Akka Persistence Query. And that, like the snapshot support, involves shared code. So in a sense it depends on this PR. So I'm glad you'll be reviewing it.

johanandren commented on src/main/scala/akka/persistence/dynamodb/journal/DynamoDBRequests.scala:
```diff
@@ -3,22 +3,90 @@
 
 trait DynamoDBRequests {
-  this: DynamoDBJournal =>
+  this: DynamoDBProvider with ActorLogging with Actor =>
```
As it is now shared between journal and snapshots I think this could be moved one package up.

Maybe just make the whole trait `private[dynamodb]` rather than individual methods

joost-de-vries replied:
Done

johanandren commented on src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotStore.scala:
```diff
@@ -0,0 +1,46 @@
+
+  override def deleteAsync(persistenceId: String, criteria: SnapshotSelectionCriteria): Future[Unit] = {
+    delete(persistenceId, criteria)
+  }
```
Just skip the curlies for these single-expression-methods

joost-de-vries replied:
Done

johanandren commented on src/main/scala/akka/persistence/dynamodb/snapshot/package.scala:
```diff
@@ -0,0 +1,81 @@
+package object snapshot {
+
+  type Item = JMap[String, AttributeValue]
+  type ItemUpdates = JMap[String, AttributeValueUpdate]
```
These are identical to the ones in the journal, put in the package object one level up?

joost-de-vries replied:
Done

johanandren commented on src/main/scala/akka/persistence/dynamodb/snapshot/package.scala:
```diff
@@ -0,0 +1,81 @@
+          new KeySchemaElement().withAttributeName(Key).withKeyType(KeyType.HASH),
+          new KeySchemaElement().withAttributeName(Timestamp).withKeyType(KeyType.RANGE)
+        ).withProjection(new Projection().withProjectionType(ProjectionType.ALL))
+    )
```
Just used in the tests as far as I can see so move it there?

joost-de-vries replied:
Done

johanandren commented on src/main/scala/akka/persistence/dynamodb/snapshot/package.scala:
```diff
@@ -0,0 +1,81 @@
+    in.foldLeft(Future.successful(cbf(in))) { (fr, a) =>
+      val fb = lift(a)
+      for (r <- fr; b <- fb) yield (r += b)
+    }.map(_.result())
```
These three methods are entirely decoupled from dynamo, but used in both stores, put in a `FutureUtils` or something like that one package up?

joost-de-vries replied:
moved to dynamodb package

johanandren commented on src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala:
```diff
@@ -0,0 +1,143 @@
+    val payloadValue = item.get(Payload).getB
+    serialization.deserialize(payloadValue.array(), classOf[Snapshot])
+      .map(snapshot =>
+        SelectedSnapshot(metadata = SnapshotMetadata(persistenceId, sequenceNr = seqNr, timestamp = timestamp), snapshot = snapshot.data)).get
```
just `.get` right away and use the result to construct the `SelectedSnapshot` instead, saves going through a lambda and a try allocation.

joost-de-vries replied:
Done

johanandren commented on src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala:
```diff
@@ -0,0 +1,143 @@
+  this: DynamoDBSnapshotStore =>
+
+  import settings._
+  import context.dispatcher
```
you could cache a `val whateverToUnit = Any => ()` up here and reuse instead of creating one for each place it is used.

In addition to that I think you could use the `akka.dispatch.ExecutionContexts.sameThreadExecutionContext` for those maps, to do that without having to jump into a potentially different thread for essentially a no-op. Maybe make that a helper method in the `FutureUtils` i mentioned somewhere else even - `discardFutureContent(Future[Any]): Future[Unit]`

joost-de-vries replied:
I've created a val as you said.

Wrt sameThreadExecutionContext: TBH I haven't used that before. What would be the proper way of using that?

ktoso replied:
Not a biggie; can get in with the normal one.
That context can be imported and used like any other one, but it hijacks the calling thread to execute the things.

johanandren commented on src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala:
```diff
@@ -0,0 +1,143 @@
+        batch => s"execute batch delete $batch",
+        result.map(snapshotDeleteReq(persistenceId, _))
+      )
+    } yield ()
```
This one is a bit messy to read, I think it would be more clear with (nicely line breaks and) map/flatMap all the way.

joost-de-vries replied:
refactored

joost-de-vries commented on README.md:
```diff
@@ -54,6 +55,29 @@ Before you can use these settings you will have to create a table, e.g. using th
+* a sort key of type Number with name `ts`
+* a local secondary index with name `ts-idx` that is an index on the combination of `par` and `ts`
+
+The Dynamodb item of a snapshot [can be 400 kB](http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Limits.html#limits-items). Using a binary serialisation format like ProtoBuf or Kryo will use that space most effectively.  
```
Do you mean: use compression in your serialisers if you're using json as a persistence format?

joost-de-vries commented on src/main/scala/akka/persistence/dynamodb/DynamoDBConfig.scala:
```diff
@@ -0,0 +1,80 @@
+  val settings: DynamoDBConfig
+  val dynamo: DynamoDBHelper
+  val serialization: Serialization
+}
```
I removed the dynamodbprovider trait. The settings I couldn't turn into a def because of the convention in this codebase of importing it, and that requires a stable value.

joost-de-vries commented on src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala:
```diff
@@ -0,0 +1,143 @@
+
+    loadQueryResult(persistenceId, criteria)
+      .map { result =>
+        val results = result.getItems.asScala.map(item => (item.get(Key).getS, item.get(SequenceNr).getN.toLong))
```
removed

joost-de-vries commented on src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala:
```diff
@@ -0,0 +1,143 @@
+        result.getItems.asScala.headOption
+          .map(youngest =>
+            fromSnapshotItem(persistenceId, youngest))
+      }
```
Done

joost-de-vries commented on src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala:
```diff
@@ -0,0 +1,143 @@
+    criteria match {
+      case SnapshotSelectionCriteria(maxSequenceNr, maxTimestamp, minSequenceNr, minTimestamp) if minSequenceNr == 0 && maxSequenceNr == Long.MaxValue =>
+        loadByTimestamp(persistenceId, minTimestamp = minTimestamp, maxTimestamp = maxTimestamp)
+      case SnapshotSelectionCriteria(maxSequenceNr, maxTimestamp, minSequenceNr, minTimestamp) if minTimestamp == 0 && maxTimestamp == Long.MaxValue =>
```
The order of those parameters are a bit unexpected. So checking their value by name makes it clearer what's going on IMO.

joost-de-vries commented on src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala:
```diff
@@ -0,0 +1,143 @@
+    item.put(Timestamp, N(timestamp))
+    val snp = B(serialization.serialize(Snapshot(snapshot)).get)
+    item.put(Payload, snp)
+    item
```
Unfortunately Dynamodbs way of determining the size is a bit complicated. So I'd rather not try to duplicate that logic.

joost-de-vries commented:
I've fixed your comments. With 2 questions.

Wrt the Dynamodb stuff:
Dynamodb is partitioned. And if you overload a single partition you get throttled. The journal implementation shards writes of a single persistenceid further to prevent write overloading.
For the journal I did not do a similar mechanism since I think typically a single snapshot at a time is made for a persistenceid.

Another design choice was how to query snapshots. The selectioncriteria can be by sequence or timestamp. I've added a local secondary index on partitionKey + timestamp. 
If the criteria search for either seq or timestamp I can read them optimally without consuming unnecessarily capacity.
I don't see how to achieve that if both sequence nr and timestamp are selected; dynamodb supports only using one index at a time.

johanandren commented:
Sharding snapshots on id sounds good.

With regards to querying, the most important part is that querying for `SnapshotSeletionCriteria.Latest` is efficient since that is what persistence normally does on every actor wakeup. I think that all the other ways of custom querying for snapshots could be slow/expensive/possibly even not implemented and just clearly documented that it will be bad to do.

joost-de-vries commented on src/main/resources/reference.conf:
```diff
@@ -121,3 +121,54 @@ dynamodb-journal {
     }
   }
 }
+dynamodb-snapshot{
```
Done

joost-de-vries commented on src/main/resources/reference.conf:
```diff
@@ -121,3 +121,54 @@ dynamodb-journal {
+
+  # If this is set to `on` then every DynamoDB request will be logged
+  # at DEBUG level. Caution: this will generate A LOT of output.
+  tracing = ${dynamodb-journal.tracing}
```
This was existing functionality & name from the journal implementation. We'll have to rename it in the journal implementation as well. With possible problems for existing users... Shall we leave it as is for now?

ktoso replied:
Yeah makes sense to keep then 👍

joost-de-vries commented on src/main/resources/reference.conf:
```diff
@@ -121,3 +121,54 @@ dynamodb-journal {
+    max-batch-write = ${dynamodb-journal.aws-api-limits.max-batch-write}
+    max-item-size = ${dynamodb-journal.aws-api-limits.max-item-size}
+  }
+}
```
Done

joost-de-vries commented on src/main/scala/akka/persistence/dynamodb/journal/DynamoDBJournalConfig.scala:
```diff
@@ -4,13 +4,13 @@
+
+class DynamoDBJournalConfig(c: Config) extends DynamoDBConfig {
   val JournalTable = c getString "journal-table"
+  val Table = JournalTable
```
fixed

joost-de-vries commented on src/test/scala/akka/persistence/dynamodb/snapshot/SnapshotStoreTckSpec.scala:
```diff
@@ -0,0 +1,25 @@
+package akka.persistence.dynamodb.snapshot
```
done

ktoso replied:
Thanks a ton :)

joost-de-vries commented on src/test/scala/akka/persistence/dynamodb/snapshot/DynamoDBUtils.scala:
```diff
@@ -0,0 +1,70 @@
+
+trait DynamoDBUtils {
+
+  val system: ActorSystem
```
done

joost-de-vries commented on src/test/scala/akka/persistence/dynamodb/snapshot/DynamoDBUtils.scala:
```diff
@@ -0,0 +1,70 @@
+    nextSeqNr += 1
+    ret
+  }
+  var generatedMessages: Vector[PersistentRepr] = Vector(null) // we start counting at 1
```
removed

joost-de-vries commented on README.md:
```diff
@@ -54,6 +55,27 @@ Before you can use these settings you will have to create a table, e.g. using th
+    journal-name =  <prefix to be used for all keys stored by this plugin>
+    aws-access-key-id =  <your key>
+    aws-secret-access-key =  <your secret>
+    endpoint =  "https://dynamodb.us-east-1.amazonaws.com" # or where your deployment is
```
Nice catch. I clarified it.

joost-de-vries commented:
@johanandren That's a good point wrt the most used query. `SnapshotSeletionCriteria.Latest` uses the min max for both seq and ts and that will be executed using the standard query by partitionkey and seq. Which is more capacity (cost) efficient than if it used the timestamp index. So that's good.

I added that the queries are using [consistent reads](http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.ReadConsistency.html). That consumes twice the capacity and fails if the network between availability zones fails. 
The journal queries use consistent reads everywhere as well.

ktoso commented:
Thanks a ton and sorry for the large delays in reviewing this - vacation season hit the team hard but we're back in pretty good capacity finally :)

I'm going to merge this and release an 1.1 so this can be used right away.

Thanks again for your persistence (pun intended) and work on this!

patriknw reviewed (commented):
👍
